### PR TITLE
upgrade to multivariatestats v0.10

### PR DESCRIPTION
**Pull request by OkonSamuel**
supersedes #45 and #49. 
This PR also
1. Adds more tests. 
2. Makes improvements to model doc strings
3. Allow users to pass `Dict` and `UnivariateFinite` objects to `priors` kwarg of `BayesianLDA` and `BayesianSubspaceLDA` models.
4. Pass adjoints to core fit method.

cc. @ablaom

### CODE REVIEW

File: src/models/discriminant_analysis.jl [hunk 1]
```diff
@@ -3,130 +3,123 @@
...
-        nclasses=nc
-    )
-    fitresult = (core_res, classes_seen)
-    return fitresult, cache, report
+  method::Symbol = :gevd::(_ in (:gevd, :whiten))
+  cov_w::CovarianceEstimator = MS.SimpleCovariance()
+  cov_b::CovarianceEstimator = MS.SimpleCovariance()
+  outdim::Int = 0::(_ ≥ 0)
+  regcoef::Float64 = 1e-6::(_ ≥ 0)
+  dist::SemiMetric = SqEuclidean()
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]: Doesn't blue style require 4 spaces for indentation? At least that's what my emacs plugin has always given me. 

```suggestion
    method::Symbol = :gevd::(_ in (:gevd, :whiten))
    cov_w::CovarianceEstimator = MS.SimpleCovariance()
    cov_b::CovarianceEstimator = MS.SimpleCovariance()
    outdim::Int = 0::(_ ≥ 0)
    regcoef::Float64 = 1e-6::(_ ≥ 0)
    dist::SemiMetric = SqEuclidean()
```

File: src/models/discriminant_analysis.jl [hunk 2]
```diff
@@ -3,130 +3,123 @@
...
-    "The number of unique classes in "*
-    "the training target has to be greater than one, even if "*
-    "the complete pool contains more than one class. If "*
-    "`cov_b=CovarianceEstimation.SimpleCovariance()` this "*
-    "restriction on classes does not apply. "
+  "The number of unique classes in "*
+  "the training target has to be greater than one, even if "*
+  "the complete pool contains more than one class. If "*
+  "`cov_b=CovarianceEstimation.SimpleCovariance()` this "*
+  "restriction on classes does not apply. "
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]:
```suggestion
    "The number of unique classes in "*
    "the training target has to be greater than one, even if "*
    "the complete pool contains more than one class. If "*
    "`cov_b=CovarianceEstimation.SimpleCovariance()` this "*
    "restriction on classes does not apply. "
```

File: src/models/discriminant_analysis.jl [hunk 3]
```diff
@@ -3,130 +3,123 @@
...
-                "`outdim` must not be larger than `min(p, nc-1)`"*
-                "where `p` is the number of features in `X` and"*
-                "`nc` is the number of unique classes in the target vector."
-            )
-        )
-    end
-    return Xm_t, yplain, classes_seen, p, n, nc, nclasses, integers_seen, outdim
+  pool = MMI.classes(y[1]) # Class list containing entries in pool of y.
+  classes_seen = unique(y) # Class list of actual entries in seen in y.
+  nc = length(classes_seen) # Number of classes in pool of y.
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]: The comment here is confusing me. Isn't the point that the `nc` is possibly *different* from the number of classes in the pool? This is number of classes seen, not the total number of classes in the pool, right?

File: src/models/discriminant_analysis.jl [hunk 4]
```diff
@@ -3,130 +3,123 @@
...
+  if outdim > def_outdim
+      throw(
+          ArgumentError(
+              "`outdim` must not be larger than `min(p, nc-1)`"*
+              "where `p` is the number of features in `X` and"*
+              "`nc` is the number of unique classes in the target vector."
+          )
+      )
+  end
+  return Xm_t, y, classes_seen, pool, p, n, nc, outdim
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]: Previously we returned `yplain = int.(y)` instead of `y` here. Is this change intentional?

File: src/models/discriminant_analysis.jl [hunk 5]
```diff
@@ -3,130 +3,123 @@
...
+end
+
+function MMI.fit(model::LDA, ::Int, X, y)
+  Xm_t, _ , classes_seen, pool, _, _, nc, outdim =
+      _check_lda_data(model, X, y)
+  core_res = MS.fit(
+      MS.MulticlassLDA, Xm_t, y;
+      method=model.method,
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]: Okay. So I guess this answers my question? MS can handle categorical vectors for `y` and the previous integer encoding was redundant? I guess there's no performance cost in *not* encoding `y`?

File: src/models/discriminant_analysis.jl [hunk 6]
```diff
@@ -3,130 +3,123 @@
...
+      Sb=MS.betweenclass_scatter(core_res),
+      Sw=MS.withclass_scatter(core_res)
+  )
+  fitresult = (core_res, classes_seen, pool)
+  return fitresult, cache, report
 end
 
-function MMI.fitted_params(::LDA, (core_res, classes_seen))
-    return (
-        projected_class_means=MS.classmeans(core_res),
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl:101]: So user can no longer access projected class means. What's the reason for no longer exposing this training outcome to the user?

File: src/models/discriminant_analysis.jl [hunk 7]
```diff
@@ -3,130 +3,123 @@
...
-
-    # compute the distances in the transformed space between pairs of rows
-    # the probability matrix Pr is `n x nc` and normalised accross rows
-    Pr = pairwise(m.dist, XWt, centroids, dims=1)
-    Pr .*= -1
-    # apply a softmax transformation
-    softmax!(Pr)
-    return MMI.UnivariateFinite(classes_seen, Pr)
+function MMI.predict(m::LDA, (core_res, classes_seen, pool), Xnew)
+  # projection of Xnew, XWt is nt x o  where o = number of out dims
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]: 4 space indentation  please

File: src/models/discriminant_analysis.jl [hunk 8]
```diff
@@ -3,130 +3,123 @@
...
 
 metadata_model(
-    LDA,
-    human_name="linear discriminant analysis model",
-    input=Table(Continuous),
-    target=AbstractVector{<:Finite},
-    weights=false,
-    output=Table(Continuous),
-    path="$(PKG).LDA"
+  LDA,
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]: And here. And everywhere else, please.

File: src/models/discriminant_analysis.jl [hunk 9]
```diff
@@ -135,329 +128,366 @@ metadata_model(
...
-        Sw=MS.withclass_scatter(core_res),
-        Sb=MS.betweenclass_scatter(core_res),
-        nclasses=nc
-    )
-
-    fitresult = (core_res, classes_seen, priors, n)
-    return fitresult, cache, report
+function _matrix_transpose(::Union{LDA,BayesianLDA}, X)
+  # MultivariateStats 9.0 is not supporting adjoints
+  return MMI.matrix(X, transpose=true)
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]: I'm pretty sure MS 0.10.0 *is* supporting adjoints. There is a separate PR for this #45. Given the many changes here, it's likely there's going to be a merge conflict. So you may want to fix that now? I think we can forget about supporting 0.9.0.

File: src/models/discriminant_analysis.jl [hunk 10]
```diff
@@ -135,329 +128,366 @@ metadata_model(
...
+
+@inline function _check_lda_priors(priors::UnivariateFinite, classes_seen, pool)
+  if MMI.classes(priors) != pool
+      throw(
+        ArgumentError("UnivariateFinite `priors` must have common pool with training target.")
+      )
+  end
+  priors_ = pdf.(priors, support(priors))
+  _check_prob01(priors_)
+  # Select priors for unique classes in `y` (For resampling purporses).
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]:
```suggestion
  # Select priors for unique classes in `y` (For resampling purposes).
```

File: src/models/discriminant_analysis.jl [hunk 11]
```diff
@@ -135,329 +128,366 @@ metadata_model(
...
+  return pdf.(priors, classes_seen)
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]: I thought we were going to make this optional checks on the CategoricalDistributions.jl side. That sounded like a sensible plan. What changed your mind?

**Review by ablaom — COMMENTED** [review]
Thanks for this @OkonSamuel .

Unfortunately all the whitespace changes (which contravene Blue Style - I checked) are making review quite tedious and I have lost steam. Can you please revert to 4-space indentation, and I will finish my review then. 

Thanks.

File: src/models/discriminant_analysis.jl (hunk 4 above)
Comment (OkonSamuel) [reply src/models/discriminant_analysis.jl]: Yes, it was. MultivariateStats directly supports categorical arrays. So it's unnecessary to call `yplain = int.(y)` first.

File: src/models/discriminant_analysis.jl (hunk 11 above)
Comment (OkonSamuel) [reply src/models/discriminant_analysis.jl]: Yes, But on giving it a second thought, we would still need to make checks again here, because there is no way of telling if the `UnivariateFinite` object the user provided has been checked during construction.

File: src/models/discriminant_analysis.jl (hunk 3 above)
Comment (OkonSamuel) [reply src/models/discriminant_analysis.jl]: Yes, `nc` is the total number of classes seen (i.e actual number of classes present in the training data) which may be different from the total  no. of classes in the pool of y. I would fix that comment as it's indeed confusing.

File: src/models/discriminant_analysis.jl (hunk 7 above)
Comment (OkonSamuel) [reply src/models/discriminant_analysis.jl]: I'm sorry about the indentation. I don't know how that got changed. I'll fix the indentation.

File: src/models/discriminant_analysis.jl (hunk 6 above)
Comment (OkonSamuel) [reply src/models/discriminant_analysis.jl:101]: It's been moved to the report. The LDA models fit parameter should be the projection matrix, all other stuff should be put in the report. Also, `MS.classmeans(core_res)` actually returns class means of the untransformed data, rather than the class means of the transformed data, so I renamed it to `class_means` rather than `projected_class_means`.

File: src/models/discriminant_analysis.jl (hunk 7 above)
Comment (OkonSamuel) [reply src/models/discriminant_analysis.jl]: It appears My VSCode indentation was somehow set to 2spaces.

File: src/models/discriminant_analysis.jl (hunk 11 above)
Comment (ablaom) [reply src/models/discriminant_analysis.jl]: Aha. I see.

**Comment by ablaom** [comment]
Sorry to complain again, but the whitespace conventions for *code* don't apply to *markdown* within docstrings. I don't know what the markdown style rules are but I would have thought there is no need to change


```
Here is list:

- `x` is a variable in 
  that is very nice

- `y` is another one
```
to:

```
- `x` is a variable in 
    that is very nice

- `y` is another one
```
Such changes are now cluttering the overall diff.

File: src/models/discriminant_analysis.jl [hunk 12]
```diff
@@ -537,34 +572,37 @@ Train the machine using `fit!(mach, rows=...)`.
...
- `indim` and `outdim` are the input and output dimensions respectively.
+  `indim` and `outdim` are the input and output dimensions respectively (See Report 
+  section below).
 
 # Report
 
 The fields of `report(mach)` are:
 
-- `classes`: The classes seen during model fitting.
+- `indim`: The dimension of the input space i.e the number of features of training matrix.
```
Comment (ablaom) [inline src/models/discriminant_analysis.jl]: Maybe "training matrix" is a little confusion as use must supply a table, right?

```suggestion
- `indim`: The dimension of the input space i.e the number of training features.
```

**Review by ablaom — APPROVED** [review]
Great to have this, thanks @OkonSamuel

File: src/models/discriminant_analysis.jl (hunk 3 above)
Comment (ablaom) [inline src/models/discriminant_analysis.jl]:
```suggestion
  nc = length(classes_seen)
```
